common_code/solver_cg_pipelined.h

```diff
@@ -0,0 +1,273 @@
+
+  {
+    if (std::is_same<PreconditionerType, dealii::PreconditionIdentity>::value == false)
+      {
+        Assert(true, ExcNotImplemented());
+      }
+
+    using number                      = typename VectorType::value_type;
+    dealii::SolverControl::State conv = dealii::SolverControl::iterate;
+
+    // Memory allocation
+    typename dealii::VectorMemory<VectorType>::Pointer z_pointer(this->memory);
+    typename dealii::VectorMemory<VectorType>::Pointer s_pointer(this->memory);
+    typename dealii::VectorMemory<VectorType>::Pointer p_pointer(this->memory);
+    typename dealii::VectorMemory<VectorType>::Pointer r_pointer(this->memory);
+    typename dealii::VectorMemory<VectorType>::Pointer w_pointer(this->memory);
+    typename dealii::VectorMemory<VectorType>::Pointer q_pointer(this->memory);
+
+    // some aliases for simpler access
+    VectorType &z = *z_pointer;
+    VectorType &s = *s_pointer;
+    VectorType &p = *p_pointer;
+    VectorType &r = *r_pointer;
+    VectorType &w = *w_pointer;
+    VectorType &q = *q_pointer;
+
+    int    it  = 0;
+    double res = std::numeric_limits<double>::max();
+
+    // resize the vectors, but do not set
+    // the values since they'd be overwritten
+    // soon anyway.
+    z.reinit(x, false);
+    s.reinit(x, false);
+    p.reinit(x, false);
+
+    r.reinit(x, true);
+    w.reinit(x, true);
+    q.reinit(x, true);
+
+    number beta      = 0.;
+    number gamma     = 0.;
+    number alpha     = 0.;
+    number gamma_old = 0.;
+    number delta     = 0.;
+
+    // compute residual. if vector is
+    // zero, then short-circuit the
+    // full computation
+    if (!x.all_zero())
+      {
+        A.vmult(r, x);
+        r.add(-1., b);
+      }
+    else
+      r.equ(-1., b);
+
+    r *= -1.0;
+
+    res = r.l2_norm();
+
+    conv = this->iteration_status(0, res, x);
+    if (conv != dealii::SolverControl::iterate)
+      return;
+
+    A.vmult(w, r);
+
+    NonBlockingDotproduct<number> nonblocking;
+
+    while (conv == dealii::SolverControl::iterate)
+      {
+        it++;
+
+        nonblocking.dot_product_start(r, r, &gamma);
+        nonblocking.dot_product_start(w, r, &delta);
+        nonblocking.dot_products_commit();
+
+        A.vmult(q, w);
+
+        nonblocking.dot_products_finish();
+
+        // compute status
+        res  = std::sqrt(gamma);
+        conv = this->iteration_status(it, res, x);
+
+        if (conv != dealii::SolverControl::iterate)
+          {
+            break;
+          }
+
+        if (it == 1)
+          {
+            alpha = gamma / delta;
+            beta  = 0.0;
+          }
+        else
+          {
+            beta  = gamma / gamma_old;
+            alpha = gamma / (delta - (beta * gamma / alpha));
+          }
+        // normal vector updates
+
+        z.sadd(beta, 1.0, q);
+        s.sadd(beta, 1.0, w);
+        p.sadd(beta, 1.0, r);
+
+        x.add(alpha, p);
+        r.add(-alpha, s);
+        w.add(-alpha, z);
```

Comment:
These could be merged to a single loop.

Reply:
I think that will be the advanced / merged version of the algorithm, won't it?

Reply:
Yes, I though for pipelined we just want the basic version.